blog: GAMMA stable (#15061)
istio/istio.io: content/en/blog/2024/gateway-mesh-ga/index.md

---

> title: "Gateway API Mesh Support Promoted To Stable"

Reviewer: Thing to take care of: we're announcing, on the Istio blog, that GW API mesh use cases are stable in GW API.  They remain Beta in Istio if feature stages has that right.

> description: The next generation traffic routing APIs are now generally available for service mesh use cases. 

Reviewer: ```suggestion
description: The next-generation Kubernetes traffic routing APIs are now Generally Available for service mesh use cases. 
```

publishdate: 2024-05-14
attribution: John Howard - solo.io
keywords: [istio, traffic, API]
target_release: 1.22
---

> We are thrilled to announce that Service Mesh support for the [Gateway API](https://gateway-api.sigs.k8s.io/) is now official "Stable"!

Reviewer: ```suggestion
We are thrilled to announce that Service Mesh support in the [Gateway API](https://gateway-api.sigs.k8s.io/) is now officially "Stable"!
```

With this release (part of Gateway API v1.1 and Istio v1.22), users can make use of the next-generation traffic management APIs for both ingress ("north-south") and service mesh use cases ("east-west").

> ## What *is* the Gateway API?

Reviewer: ```suggestion
## What is the Gateway API?
```

Reviewer: Could you _be_ any more gateway?

The Gateway API is a collection of APIs that are part of Kubernetes, focusing on traffic routing and management.

> The APIs are inspired by, and serve many of the same roles as, Kubernetes' `Ingress`, and Istio's `VirtualService` and `Gateway` APIs.

Reviewer: ```suggestion
The APIs are inspired by, and serve many of the same roles as, Kubernetes' `Ingress` and Istio's `VirtualService` and `Gateway` APIs.
```

These APIs have been under development both in Istio, as well as with [broad collaboration](https://gateway-api.sigs.k8s.io/implementations/), since 2020, and have come a long way since then.

> While the API initially targeted only serving ingress use cases (which went GA [last year](https://kubernetes.io/blog/2023/10/31/gateway-api-ga/), we had always envisioned allowing the same APIs to be used for traffic *within* a cluster as well.

Reviewer: ```suggestion
While the API initially targeted only serving ingress use cases (which went GA [last year](https://kubernetes.io/blog/2023/10/31/gateway-api-ga/)), we had always envisioned allowing the same APIs to be used for traffic *within* a cluster as well.
```

With this release, that vision is made a reality: Istio users can use the same routing API for all of their traffic!

## Getting started

Throughout the Istio documentation, all of our examples have been updated to show how to use the Gateway API, so explore some of the [tasks](/docs/tasks/traffic-management/) to gain a deeper understanding.

Using Gateway API for service mesh should feel familiar both to users already using Gateway API for ingress, and users using `VirtualService` for service mesh today.

> * Compared to Gateway API for ingress, the routes target a `Service` instead of a `Gateway`.

Reviewer: ```suggestion
* Compared to Gateway API for ingress, routes target a `Service` instead of a `Gateway`.
```

> * Compared to VirtualService, where routes associate with a set of `hosts`, routes target a `Service`.

Reviewer: standard seems to be "virtual services" or `VirtualService`

> * Compared to `VirtualService`, where routes associate with a set of `hosts`, routes target a `Service`.

Reviewer 2: Maybe at a warning here about Cluster IP vs. host/authority header routing?

Reviewer 2: Oh wait sorry I'm thinking about Ambient; yeah the warning about not implementing GAMMA all the way is good too

> Here is a simple example, which showcases routing requests to two different versions of a `Service` based on the request.

Reviewer: ```suggestion
Here is a simple example, which demonstrates routing requests to two different versions of a `Service` based on the request header:
```

{{< text yaml >}}
apiVersion: gateway.networking.k8s.io/v1
kind: HTTPRoute
metadata:
  name: reviews
spec:
  parentRefs:
  - group: ""
    kind: Service
    name: reviews
    port: 9080
  rules:
  - matches:
    - headers:
      - name: my-favorite-service-mesh
        value: istio
    filters:
    - type: RequestHeaderModifier
      requestHeaderModifier:
      add:
        - name: hello
          value: world
    backendRefs:
    - name: reviews-v2
      port: 9080
  - backendRefs:
    - name: reviews-v1
      port: 9080
{{< /text >}}

Breaking this down, we have a few parts:
* First, we identify what routes we should match.

>   By attaching our route to the `reviews` `Service`, we will apply this routing configuration to all requests that were originally targeting `reviews`.

Reviewer: ```suggestion
  By attaching our route to the `reviews` Service, we will apply this routing configuration to all requests that were originally targeting `reviews`.
```

(largely so it doesn't look like one term)

> * Next, we can (optionally) have a variety of matches on the request if we want to behave differently based on the request. This is handled by the `matches`.

Reviewer: ```suggestion
* Next, we can (optionally) have a variety of matches on the request, if we want to behave differently based on the request. This is handled by the `matches`.
```

(Can you have a go at rewriting this?  It doesn't make sense/repeats 'matches' and 'request' too much for my liking)

* Optionally, we can modify the request. Here, we add a header.
* Finally, we select a destination for the request. In this example, we are picking between two versions of our application.

For more details, see [Istio's traffic routing internals](/docs/ops/configuration/traffic-management/traffic-routing/) and [Gateway API's Service documentation](https://gateway-api.sigs.k8s.io/mesh/service-facets/).

## Which API should I use?

With overlapping responsibilities (and names!), picking which APIs to use can be a bit confusing.

Here is the breakdown:

| API Name     | Object Types                                                                                                                          | Status                            | Recommendation                                                             |
|--------------|---------------------------------------------------------------------------------------------------------------------------------------|-----------------------------------|----------------------------------------------------------------------------|

> | Gateway APIs | [HTTPRoute](https://gateway-api.sigs.k8s.io/api-types/httproute/), [Gateway](https://gateway-api.sigs.k8s.io/api-types/gateway/), ... | Stable in Gateway API v1.0 (2023) | Use for new deployments, in particular with [ambient mode](/docs/ambient/) |

Reviewer 3: Is Gateway API for sidecar stable but for ambient remains beta as ambient is still beta?

Author: The stability of **any** feature is the `min(intersection of features used)`. For exampe, I think we claim multi-clsuter in istio is beta (whether it should be is out of scope for this comment :slightly_smiling_face: ).

We don't need to say "Gateway API for Multiclsuter is beta" and "Gateway API for dual stack is experimental" and "Gateway API for ambient is beta".

Reviewer 3: Understood, just thought users would ask that question as they read the blog as ambient mode is mentioned here.

> | Istio APIs   | [Virtual Service](/docs/reference/config/networking/virtual-service/), [Gateway](/docs/reference/config/networking/gateway/)          | `v1` in Istio 1.22 (2024)         | Use for existing deployments, or where advanced features are needed        |
> | Gateway APIs | [HTTPRoute](https://gateway-api.sigs.k8s.io/api-types/httproute/), [Gateway](https://gateway-api.sigs.k8s.io/api-types/gateway/), ... | Stable in Gateway API v1.0 (2023) | Use for new deployments, in particular with [ambient mode](/docs/ambient/) |

Reviewer: ```suggestion
| Gateway APIs | [HTTPRoute](https://gateway-api.sigs.k8s.io/api-types/httproute/), [Gateway](https://gateway-api.sigs.k8s.io/api-types/gateway/), ... | Stable in Gateway API v1.0 (2023) | Use for new deployments, in particular with [ambient mode](/docs/ambient/) |
| Istio APIs   | [Virtual Service](/docs/reference/config/networking/virtual-service/), [Gateway](/docs/reference/config/networking/gateway/)          | `v1` in Istio 1.22 (2024)         | Use for existing deployments, or where advanced features are needed        |
```

switching order to go from best to worst, as it were

| Ingress API  | [Ingress](https://kubernetes.io/docs/concepts/services-networking/ingress)                                                            | Stable in Kubernetes v1.19 (2020) | Use only for legacy deployments                                            |

> You may wonder, given the above, why were Istio APIs [promoted to `v1`](/blog/2024/v1-apis) concurrently?

Reviewer: ```suggestion
You may wonder, given the above, why the Istio APIs were [promoted to `v1`](/blog/2024/v1-apis) concurrently?
```

> This was part of an effort to accurate categorize the *stability* of the APIs.

Reviewer 3: Is there anything here should be tied to the release channel stuff? cc @whitneygriffith

While we view Gateway API as the future (and present!) of traffic routing APIs, our existing APIs are here to stay for the long run, with full compatibility.

> This mirrors Kubernete's approach with [`Ingress`](https://kubernetes.io/docs/concepts/services-networking/ingress), which was promoted to `v1` while directing future work towards the Gateway API.

Reviewer: ```suggestion
This mirrors Kubernetes' approach with [`Ingress`](https://kubernetes.io/docs/concepts/services-networking/ingress), which was promoted to `v1` while directing future work towards the Gateway API.
```

## Community

> This stability graduation represents the culmination of countless hours of work and collaboration across an incredible diverse group.

Reviewer: ```suggestion
This stability graduation represents the culmination of countless hours of work and collaboration across an incredible, diverse group.
```

or incredibly diverse group.
Though the word 'diverse' has a somewhat more narrow meaning these days and as such you might like to avoid it entirely.  I would have replaced it with a suitable synonym had one come to mind

Author: Cleaned it up a bit, I think it reads better now

> It is incredible to look at the [diverse list of organizations](https://gateway-api.sigs.k8s.io/implementations/) involved in the API and look back at how far we have come.

Reviewer: ```suggestion
It is incredible to look at the [list of organizations](https://gateway-api.sigs.k8s.io/implementations/) involved in the API and consider how far we have come.
```

(remove repeated words)

Author: :+1:

> A special thanks goes out to my [co-leads on the effort](https://gateway-api.sigs.k8s.io/mesh/gamma/): Flynn, Keith Mattix, and Mike Morris, as well as the countless others involved.

Reviewer: ```suggestion
A special thanks goes out to [my co-leads on the effort](https://gateway-api.sigs.k8s.io/mesh/gamma/): Flynn, Keith Mattix, and Mike Morris, as well as the countless others involved.
```

nit: give Flynn another name, please

Author: It is my understanding that is their chosen representation (see https://kubernetes.io/blog/2024/05/09/gateway-api-v1-1/, their github, etc). I can double check with them.

Interested in getting involved, or even just providing feedback?

> Check out our [Community page](/get-involved/) or the Gateway API [contributing guide](https://gateway-api.sigs.k8s.io/contributing/)!

Reviewer: ```suggestion
Check out Istio's [community page](/get-involved/) or the Gateway API [contributing guide](https://gateway-api.sigs.k8s.io/contributing/)!

```
